src/main/flight/position.c

```diff
@@ -22,204 +22,202 @@
+static bool wasArmed = false;
+static bool isGpsAvailable = false;     // true if GPS is connected and while it has a 3D fix, set each run to false
+static bool isBaroAvailable = false;    // true if baro exists and has been calibrated on power up
+
+static float gpsAltCm = 0.0f;           // will hold last value on transient loss of 3D fix
+static float gpsAltGroundCm = 0.0f;
+static float baroAltCm = 0.0f;
+
+#if defined(USE_GPS) && defined(USE_BARO)
+static pt1Filter_t gpsFuseLpf;          // for complementary filter (lowpass part)
+static pt1Filter_t baroFuseLpf;         // for complementary filter (highpass part)
+static float gpsAltLowpassCm = 0.0f;
+static float baroAltHighpassCm = 0.0f;
 #endif
 
-static pt2Filter_t altitudeLpf;
-static pt2Filter_t altitudeDerivativeLpf;
 #ifdef USE_VARIO
-static int16_t estimatedVario = 0; // in cm/s
+static pt2Filter_t varioLpf;
+static float climbRateCmS = 0.0f;
+static int16_t estimatedVario = 0;      // in cm/s
 #endif
 
 void positionInit(void)
 {
     const float sampleTimeS = HZ_TO_INTERVAL(TASK_ALTITUDE_RATE_HZ);
 
-    const float altitudeCutoffHz = positionConfig()->altitude_lpf / 100.0f;
-    const float altitudeGain = pt2FilterGain(altitudeCutoffHz, sampleTimeS);
-    pt2FilterInit(&altitudeLpf, altitudeGain);
+    float cutoffHz;
+    float gain;
 
-    const float altitudeDerivativeCutoffHz = positionConfig()->altitude_d_lpf / 100.0f;
-    const float altitudeDerivativeGain = pt2FilterGain(altitudeDerivativeCutoffHz, sampleTimeS);
-    pt2FilterInit(&altitudeDerivativeLpf, altitudeDerivativeGain);
-}
+#ifdef USE_GPS
+    // initialize GPS upsampling filter
+    const float gpsRateHz = 1.0f / getGpsDataIntervalSeconds();
+    cutoffHz = gpsRateHz / 4.0f;    // -3dB cutoff at nyquist/2
+    gain = pt2FilterGain(cutoffHz, sampleTimeS);
+    pt2FilterInit(&gpsUpsampleLpf, gain);
+#endif
 
-typedef enum {
-    DEFAULT = 0,
-    BARO_ONLY,
-    GPS_ONLY
-} altitudeSource_e;
+#ifdef USE_BARO
+    // initialize baro upsampling filter
+    const float baroRateHz = 1.0f / ((baro.dev.up_delay + 1000 * (BARO_STATE_COUNT - 1)) * 1e-6f);  // up_delay + 5x 1000us baro state delay
+    cutoffHz = baroRateHz / 4.0f;   // -3dB cutoff at nyquist/2
+    gain = pt2FilterGain(cutoffHz, sampleTimeS);
+    pt2FilterInit(&baroUpsampleLpf, gain);
+#endif
 
-PG_REGISTER_WITH_RESET_TEMPLATE(positionConfig_t, positionConfig, PG_POSITION, 4);
+#if defined(USE_GPS) && defined(USE_BARO)
+    // initialize complementary filter for fusing GPS and barometer
+    cutoffHz = gpsRateHz / 2.0f * positionConfig()->altitude_fuse_ratio / 100.0f;
+    gain = pt1FilterGain(cutoffHz, sampleTimeS);
+    pt1FilterInit(&baroFuseLpf, gain);
+    pt1FilterInit(&gpsFuseLpf, gain);
+#endif
 
-PG_RESET_TEMPLATE(positionConfig_t, positionConfig,
-    .altitude_source = DEFAULT,
-    .altitude_prefer_baro = 100, // percentage 'trust' of baro data
-    .altitude_lpf = 300,
-    .altitude_d_lpf = 100,
-);
+#ifdef USE_VARIO
```

Comment:
Currently USE_VARIO is automatically defined if USE_BARO or USE_GYRO are defined. [common_post.h]